### PR TITLE
dbt-materialize: consider v2 of `source` and `sink` materializations

### LINKED ISSUE #17313: dbt-materialize: add `cluster` config to source and sink materialization logic
### Feature request

Materialize v0.41.0 will ship with the new cluster unification logic ([Cloud #4929](https://github.com/MaterializeInc/cloud/issues/4929)), which allows users to assign sources and sinks to a specified cluster. This requires us to allow the [`cluster` config option](https://materialize.com/docs/integrations/dbt/#clusters) in the `dbt-materialize` adapter also for the source and sink materializations.

## TODO
- [ ] Change the logic in the adapter
- [ ] Update the [dbt integration guide](https://materialize.com/docs/integrations/dbt/#configuration)
- [ ] Update the [dbt documentation](https://docs.getdbt.com/reference/resource-configs/materialize-configs)

### PR DESCRIPTION
**Pull request by morsapaes**
Fixes #17313.

### CODE REVIEW

File: misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql [hunk 1]
```diff
@@ -86,6 +86,40 @@
...
+  {% if contract_config.enforced %}
+    {{exceptions.warn("Model contracts cannot be enforced for custom materializations (see dbt-core #7213)")}}
+  {%- endif %}
+
+  {%- set cluster = config.get('cluster', target.cluster) -%}
+
+  create source {{ relation }}
+  {% if cluster %}
+    in cluster {{ cluster }}
+  {% endif %}
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:100]: I was thinking that we probably want to include the `FROM` here too, for consistency with `CREATE VIEW ... AS`. I.e., the preposition is stripped off in both cases.

```suggestion
  {% endif %}
from
```

File: misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql [hunk 2]
```diff
@@ -86,6 +86,40 @@
...
+  {% if contract_config.enforced %}
+    {{exceptions.warn("Model contracts cannot be enforced for custom materializations (see dbt-core #7213)")}}
+  {%- endif %}
+  
+  {%- set cluster = config.get('cluster', target.cluster) -%}
+
+  create sink {{ relation }}
+  {% if cluster %}
+    in cluster {{ cluster }}
+  {% endif %}
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:117]: ...oh, but that approach would be awkward for sinks. 😞

**Review by benesch — COMMENTED** [review]
I like it!

File: misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql (hunk 2 above)
Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:117]: 🫠

Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:117]: I mean...adding it to the `source` materialization but not the sink one would make the DDL users write consistent, no?

Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:117]: Ah, nevermind.

**Comment by sjwiesman** [comment]
```patch
diff --git a/misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql b/misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql
index 30d14eb78f..37b62d04b8 100644
--- a/misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql
+++ b/misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql
@@ -92,7 +92,7 @@
     {{exceptions.warn("Model contracts cannot be enforced for custom materializations (see dbt-core #7213)")}}
   {%- endif %}
 
-  {%- set cluster = config.get('cluster', target.cluster) -%}
+  {%- set cluster = generate_cluster_name(config.get('cluster', target.cluster)) -%}
 
   create source {{ relation }}
   {% if cluster %}
@@ -109,7 +109,7 @@
     {{exceptions.warn("Model contracts cannot be enforced for custom materializations (see dbt-core #7213)")}}
   {%- endif %}
 
-  {%- set cluster = config.get('cluster', target.cluster) -%}
+  {%- set cluster = generate_cluster_name(config.get('cluster', target.cluster)) -%}
 
   create sink {{ relation }}
   {% if cluster %}
diff --git a/misc/dbt-materialize/dbt/include/materialize/macros/deploy/deploy_promote.sql b/misc/dbt-materialize/dbt/include/materialize/macros/deploy/deploy_promote.sql
index 4b49df4ae7..8380f99c7c 100644
--- a/misc/dbt-materialize/dbt/include/materialize/macros/deploy/deploy_promote.sql
+++ b/misc/dbt-materialize/dbt/include/materialize/macros/deploy/deploy_promote.sql
@@ -56,9 +56,9 @@
     {% if not schema_exists(deploy_schema) %}
         {{ exceptions.raise_compiler_error("Deployment schema " ~ deploy_schema ~ " does not exist") }}
     {% endif %}
-    {% if schema_contains_sources_or_sinks(deploy_schema) %}
+    {% if schema_contains_sinks(deploy_schema) %}
         {{ exceptions.raise_compiler_error("""
-        Deployment schema " ~ deploy_schema ~ " contains sources or sinks. This is not currently
+        Deployment schema " ~ deploy_schema ~ " contains sinks. This is not currently
         supported by dbt run-operation deploy_promote.
 
         If this feature is important to you, please reach out!
@@ -74,9 +74,9 @@
     {% if not cluster_exists(deploy_cluster) %}
         {{ exceptions.raise_compiler_error("Deployment cluster " ~ deploy_cluster ~ " does not exist") }}
     {% endif %}
-    {% if cluster_contains_sources_or_sinks(deploy_cluster) %}
+    {% if cluster_contains_sinks(deploy_cluster) %}
         {{ exceptions.raise_compiler_error("""
-        Deployment cluster " ~ deploy_cluster ~ " contains sources or sinks. This is not currently
+        Deployment cluster " ~ deploy_cluster ~ " contains sinks. This is not currently
         supported by dbt run-operation deploy_promote.
 
         If this feature is important to you, please reach out!
@@ -107,21 +107,11 @@ COMMIT;
 {%- endcall %}
 {% endmacro %}
 
-{% macro cluster_contains_sources_or_sinks(cluster) %}
+{% macro cluster_contains_sinks(cluster) %}
     {% set query %}
-        WITH sources_and_sinks AS (
-            SELECT cluster_id
-            FROM mz_sources
-
-            UNION ALL
-
-            SELECT cluster_id
-            FROM mz_sinks
-        )
-
         SELECT count(*) > 0
-        FROM sources_and_sinks
-        JOIN mz_clusters ON sources_and_sinks.cluster_id = mz_clusters.id
+        FROM mz_sinks
+        JOIN mz_clusters ON mz_sinks.cluster_id = mz_clusters.id
         WHERE mz_clusters.name = {{ dbt.string_literal(cluster) }}
     {% endset %}
 
@@ -131,21 +121,11 @@ COMMIT;
     {% endif %}
 {% endmacro %}
 
-{% macro schema_contains_sources_or_sinks(schema) %}
+{% macro schema_contains_sinks(schema) %}
     {% set query %}
-        WITH sources_and_sinks AS (
-            SELECT schema_id
-            FROM mz_sources
-
-            UNION ALL
-
-            SELECT schema_id
-            FROM mz_sinks
-        )
-
         SELECT count(*) > 0
-        FROM sources_and_sinks
-        JOIN mz_schemas ON sources_and_sinks.schema_id = mz_schemas.id
+        FROM mz_sinks
+        JOIN mz_schemas ON mz_sinks.schema_id = mz_schemas.id
         JOIN mz_databases ON mz_schemas.database_id = mz_databases.id
         WHERE mz_schemas.name = {{ dbt.string_literal(schema) }}
             AND mz_databases.name = current_database()

```

If you apply this patch, it should enable blue / green for sources

**Review by sjwiesman — APPROVED** [review]

**Comment by sjwiesman** [comment]
:itshappeningmeme:

**Comment by morsapaes** [comment]
There's some documentation work to do as a follow-up, I'll cut a release once that's through.

**Pull request merged**